crates/ruff_python_formatter/src/comments/format.rs

```diff
@@ -146,6 +147,56 @@ impl Format<PyFormatContext<'_>> for FormatDanglingComments<'_> {
     }
 }
 
+/// Formats the content of the passed comment.
+///
+/// * Adds a whitespace between `#` and the comment text except if the first character is a `#`, `:`, `'`, or `!`
+/// * Replaces non breaking whitespaces with regular whitespaces except if in front of a `types:` comment
+const fn format_comment(comment: &SourceComment) -> FormatComment {
+    FormatComment { comment }
+}
+
+struct FormatComment<'a> {
+    comment: &'a SourceComment,
+}
+
+impl Format<PyFormatContext<'_>> for FormatComment<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        let slice = self.comment.slice();
+        let comment_text = slice.text(SourceCode::new(f.context().contents()));
+
+        let (content, mut start_offset) = comment_text
+            .strip_prefix('#')
+            .map_or((comment_text, TextSize::new(0)), |rest| {
+                (rest, '#'.text_len())
+            });
+
+        write!(f, [source_position(slice.start()), text("#")])?;
+
+        // Starts with a non breaking space
+        if content.starts_with('\u{A0}') && !content.trim_start().starts_with("type:") {
+            // Replace non-breaking space with a space (if not followed by a normal space)
+            start_offset += '\u{A0}'.text_len();
+        }
+
+        // Add a space between the `#` and the text if the source contains none.
+        if !content.is_empty() && !content.starts_with([' ', '!', ':', '#', '\'']) {
+            write!(f, [space()])?;
+        }
+
+        let start = slice.start() + start_offset;
+        let trimmed = content.trim_end();
+        let end = slice.range().end() - (content.text_len() - trimmed.text_len());
```

Comment:
found it really hard to understand that here we just want to remove [the length of] the trimmed end

Reply:
I introduced a temporary variable `trailing_whitespace_len`. I hope that will make this clearer